File: tests/conformance/conformance.test.ts
/**
 * pdftract SDK Conformance Test Runner (Node.js / TypeScript)
 *
 * This test runs the shared SDK conformance suite against the Node.js SDK.
 * It loads tests/sdk-conformance/cases.json and executes each test case.
 *
 * Run with: vitest test/conformance/conformance.test.ts
 * Or as standalone: ts-node test/conformance/conformance.test.ts
 */

import { readFileSync, writeFileSync } from 'fs';
import { join } from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = join(__filename, '..');

const SUITE_PATH = join(__dirname, '..', '..', 'sdk-conformance', 'cases.json');
const SDK_NAME = 'pdftract-node';
const SDK_VERSION = '0.1.0';

enum TestStatus {
  Pass = 'pass',
  Fail = 'fail',
  Skip = 'skip',
  Error = 'error',
}

interface TestResult {
  id: string;
  status: TestStatus;
  actual?: any;
  expected?: any;
  error?: string;
  reason?: string;
  duration_ms: number;
}

interface ConformanceReport {
  sdk: string;
  sdk_version: string;
  suite_version: string;
  schema_version: string;
  timestamp: string;
  results: TestResult[];
  summary: {
    total: number;
    passed: number;
    failed: number;
    skipped: number;
    errors: number;
    duration_ms: number;
  };
  environment: {
    os: string;
    arch: string;
    binary_version: string;
    runtime_version: string;
  };
}

interface SuiteCase {
  id: string;
  fixture: string;
  method: string;
  options: Record<string, any>;
  expected: any;
  tolerances?: Record<string, { abs?: number; rel?: number }>;
  feature?: string;
  min_schema_version?: string;
  skip_reason?: string;
}

interface Suite {
  version: string;
  schema_version: string;
  cases: SuiteCase[];
}

function loadSuite(path: string): Suite {
  const content = readFileSync(path, 'utf-8');
  return JSON.parse(content);
}

function compareWithTolerance(
  actual: number,
  expected: number,
  tolerance?: { abs?: number; rel?: number }
): boolean {
  if (!tolerance) {
    return Math.abs(actual - expected) < Number.EPSILON;
  }

  if (tolerance.abs !== undefined) {
    if (Math.abs(actual - expected) <= tolerance.abs) {
      return true;
    }
  }

  if (tolerance.rel !== undefined) {
    const diff = Math.abs(actual - expected);
    const avg = (actual + expected) / 2.0;
    if (avg > 0.0 && diff / avg <= tolerance.rel) {
      return true;
    }
  }

  return false;
}

function findTolerance(
  tolerances: Record<string, any> | undefined,
  path: string
): { abs?: number; rel?: number } | undefined {
  if (!tolerances) {
    return undefined;
  }

  if (path in tolerances) {
    return tolerances[path];
  }

  for (const [key, val] of Object.entries(tolerances)) {
    if (key.includes('*')) {
      const pattern = key.replace(/\*/g, '.*');
      const regex = new RegExp(pattern);
      if (regex.test(path)) {
        return val;
      }
    }
  }

  return undefined;
}

function compareResults(
  actual: any,
  expected: any,
  tolerances: Record<string, any> | undefined,
  path: string = ''
): { passed: boolean; reason?: string } {
  if (typeof expected === 'object' && expected !== null && !Array.isArray(expected)) {
    if ('min' in expected && typeof actual === 'number') {
      if (actual < expected.min) {
        return { passed: false, reason: `${path}: value ${actual} < minimum ${expected.min}` };
      }
    }
    if ('max' in expected && typeof actual === 'number') {
      if (actual > expected.max) {
        return { passed: false, reason: `${path}: value ${actual} > maximum ${expected.max}` };
      }
    }
    if ('value' in expected && typeof actual === 'number') {
      const tol = findTolerance(tolerances, path);
      if (!compareWithTolerance(actual, expected.value, tol)) {
        return { passed: false, reason: `${path}: numeric mismatch` };
      }
    }
    if ('min_length' in expected && typeof actual === 'string') {
      if (actual.length < expected.min_length) {
        return { passed: false, reason: `${path}: string length ${actual.length} < minimum ${expected.min_length}` };
      }
    }
    if ('contains' in expected && typeof actual === 'string') {
      for (const substring of expected.contains) {
        if (!actual.includes(substring)) {
          return { passed: false, reason: `${path}: string does not contain '${substring}'` };
        }
      }
    }
    if ('min' in expected && Array.isArray(actual)) {
      if (actual.length < expected.min) {
        return { passed: false, reason: `${path}: array length ${actual.length} < minimum ${expected.min}` };
      }
    }
    if ('max' in expected && Array.isArray(actual)) {
      if (actual.length > expected.max) {
        return { passed: false, reason: `${path}: array length ${actual.length} > maximum ${expected.max}` };
      }
    }

    // Nested object comparison
    if (typeof actual === 'object' && actual !== null) {
      for (const [key, expVal] of Object.entries(expected)) {
        const newPath = path ? `${path}.${key}` : key;
        if (!(key in actual)) {
          return { passed: false, reason: `${newPath}: missing key '${key}'` };
        }
        const result = compareResults(actual[key], expVal, tolerances, newPath);
        if (!result.passed) {
          return result;
        }
      }
    }
  } else if (Array.isArray(expected) && Array.isArray(actual)) {
    for (let i = 0; i < expected.length; i++) {
      const newPath = `${path}[${i}]`;
      if (i >= actual.length) {
        return { passed: false, reason: `${newPath}: missing index` };
      }
      const result = compareResults(actual[i], expected[i], tolerances, newPath);
      if (!result.passed) {
        return result;
      }
    }
  } else {
    if (actual !== expected) {
      return { passed: false, reason: `${path}: expected ${JSON.stringify(expected)}, got ${JSON.stringify(actual)}` };
    }
  }

  return { passed: true };
}

async function executeMethod(
  method: string,
  fixture: string,
  options: Record<string, any>
): Promise<any> {
  // This is a stub - replace with actual SDK calls when available
  switch (method) {
    case 'extract':
      return {
        schema_version: '1.0',
        metadata: { page_count: 1 },
        pages: [
          {
            page_index: 0,
            width: 612,
            height: 792,
            rotation: 0,
          },
        ],
        errors: [],
      };
    case 'extract_text':
      return 'Sample text content';
    case 'extract_markdown':
      return '# Sample Markdown\n\nContent here';
    case 'extract_stream':
      return { output_type: 'iterator', frame_count: 3 };
    case 'search':
      return { output_type: 'iterator', matches: [{ page: 0, text: 'found' }] };
    case 'get_metadata':
      return { metadata: { page_count: 1, title: 'Test', author: 'Test' } };
    case 'hash':
      return { hash: 'abc123', fast_hash: 'def456' };
    case 'classify':
      return { category: 'scientific_paper', confidence: 0.85, tags: ['academic'] };
    case 'verify_receipt':
      return { valid: true };
    default:
      return null;
  }
}

async function runTestCase(
  case: SuiteCase,
  schemaVersion: string,
  fixturesBase: string
): Promise<TestResult> {
  const startTime = Date.now();

  // Check min_schema_version
  if (case.min_schema_version) {
    const [major, minor] = schemaVersion.split('.').map(Number);
    const [minMajor, minMinor] = case.min_schema_version.split('.').map(Number);
    if (major < minMajor || (major === minMajor && minor < minMinor)) {
      return {
        id: case.id,
        status: TestStatus.Skip,
        reason: `Schema version ${schemaVersion} < minimum required ${case.min_schema_version}`,
        duration_ms: Date.now() - startTime,
      };
    }
  }

  const fixturePath = case.fixture.startsWith('http')
    ? case.fixture
    : join(fixturesBase, case.fixture);

  try {
    const actual = await executeMethod(case.method, fixturePath, case.options);
    const { passed, reason } = compareResults(actual, case.expected, case.tolerances);

    return {
      id: case.id,
      status: passed ? TestStatus.Pass : TestStatus.Fail,
      actual,
      expected: case.expected,
      reason,
      duration_ms: Date.now() - startTime,
    };
  } catch (e) {
    return {
      id: case.id,
      status: TestStatus.Error,
      expected: case.expected,
      error: e instanceof Error ? e.message : String(e),
      duration_ms: Date.now() - startTime,
    };
  }
}

export async function runConformance(
  suitePath: string = SUITE_PATH,
  outputPath: string = 'conformance-report.json'
): Promise<ConformanceReport> {
  const os = process.platform;
  const arch = process.arch;
  const runtimeVersion = `Node.js ${process.version}`;

  console.log(`pdftract SDK Conformance Runner`);
  console.log(`SDK: ${SDK_NAME} v${SDK_VERSION}`);
  console.log(`Suite: ${suitePath}`);
  console.log();

  const suite = loadSuite(suitePath);
  const fixturesBase = join(suitePath, '..', 'fixtures');

  console.log(`Found ${suite.cases.length} test cases`);
  console.log();

  const startTime = Date.now();
  const results: TestResult[] = [];

  for (const case_ of suite.cases) {
    const result = await runTestCase(case_, suite.schema_version, fixturesBase);
    const statusSym = {
      [TestStatus.Pass]: 'PASS',
      [TestStatus.Fail]: 'FAIL',
      [TestStatus.Skip]: 'SKIP',
      [TestStatus.Error]: 'ERROR',
    }[result.status];

    console.log(`[${statusSym}] ${result.id} (${result.duration_ms}ms)`);

    if (result.status === TestStatus.Fail || result.status === TestStatus.Error) {
      if (result.reason) {
        console.log(`  Reason: ${result.reason}`);
      }
      if (result.error) {
        console.log(`  Error: ${result.error}`);
      }
    }

    results.push(result);
  }

  const duration_ms = Date.now() - startTime;

  const summary = {
    total: results.length,
    passed: results.filter((r) => r.status === TestStatus.Pass).length,
    failed: results.filter((r) => r.status === TestStatus.Fail).length,
    skipped: results.filter((r) => r.status === TestStatus.Skip).length,
    errors: results.filter((r) => r.status === TestStatus.Error).length,
    duration_ms,
  };

  console.log();
  console.log('Summary:');
  console.log(`  Total:   ${summary.total}`);
  console.log(`  Passed:  ${summary.passed}`);
  console.log(`  Failed:  ${summary.failed}`);
  console.log(`  Skipped: ${summary.skipped}`);
  console.log(`  Errors:  ${summary.errors}`);
  console.log(`  Time:    ${summary.duration_ms}ms`);

  const report: ConformanceReport = {
    sdk: SDK_NAME,
    sdk_version: SDK_VERSION,
    suite_version: suite.version,
    schema_version: suite.schema_version,
    timestamp: new Date().toISOString(),
    results,
    summary,
    environment: {
      os,
      arch,
      binary_version: SDK_VERSION,
      runtime_version: runtimeVersion,
    },
  };

  writeFileSync(outputPath, JSON.stringify(report, null, 2));
  console.log();
  console.log(`Report written to: ${outputPath}`);

  return report;
}

// Vitest entry point
export async function testConformanceSuite() {
  const report = await runConformance();
  if (report.summary.failed > 0) {
    throw new Error(`${report.summary.failed} tests failed`);
  }
  if (report.summary.errors > 0) {
    throw new Error(`${report.summary.errors} tests errored`);
  }
}

// CLI entry point
if (import.meta.url === `file://${process.argv[1]}`) {
  const suiteArg = process.argv[2];
  const outputArg = process.argv[3];

  runConformance(suiteArg, outputArg).then((report) => {
    process.exit(report.summary.failed === 0 && report.summary.errors === 0 ? 0 : 1);
  });
}

File: tests/conformance/ConformanceTest.java
/**
 * pdftract SDK Conformance Test Runner (Java)
 *
 * This test runs the shared SDK conformance suite against the Java SDK.
 * It loads tests/sdk-conformance/cases.json and executes each test case.
 *
 * Run with: mvn test -Dtest=ConformanceTest
 * Or as standalone: java ConformanceTest.java <suite-path> <output-path>
 */

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.time.Instant;
import java.time.ZoneOffset;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

public class ConformanceTest {

    private static final String SUITE_PATH = "tests/sdk-conformance/cases.json";
    private static final String SDK_NAME = "pdftract-java";
    private static final String SDK_VERSION = "0.1.0";

    private final ObjectMapper mapper = new ObjectMapper();

    enum TestStatus {
        PASS, FAIL, SKIP, ERROR
    }

    static class TestResult {
        String id;
        TestStatus status;
        JsonNode actual;
        JsonNode expected;
        String error;
        String reason;
        long durationMs;

        TestResult(String id, TestStatus status, long durationMs) {
            this.id = id;
            this.status = status;
            this.durationMs = durationMs;
        }
    }

    static class ConformanceReport {
        String sdk;
        String sdkVersion;
        String suiteVersion;
        String schemaVersion;
        String timestamp;
        List<TestResult> results;
        Summary summary;
        Environment environment;

        ObjectNode toJson(ObjectMapper mapper) {
            ObjectNode node = mapper.createObjectNode();
            node.put("sdk", sdk);
            node.put("sdk_version", sdkVersion);
            node.put("suite_version", suiteVersion);
            node.put("schema_version", schemaVersion);
            node.put("timestamp", timestamp);

            var resultsArray = node.putArray("results");
            for (var result : results) {
                var resultNode = resultsArray.addObject();
                resultNode.put("id", result.id);
                resultNode.put("status", result.status.name().toLowerCase());
                if (result.actual != null) {
                    resultNode.set("actual", result.actual);
                }
                if (result.expected != null) {
                    resultNode.set("expected", result.expected);
                }
                if (result.error != null) {
                    resultNode.put("error", result.error);
                }
                if (result.reason != null) {
                    resultNode.put("reason", result.reason);
                }
                resultNode.put("duration_ms", result.durationMs);
            }

            var summaryNode = node.putObject("summary");
            summaryNode.put("total", summary.total);
            summaryNode.put("passed", summary.passed);
            summaryNode.put("failed", summary.failed);
            summaryNode.put("skipped", summary.skipped);
            summaryNode.put("errors", summary.errors);
            summaryNode.put("duration_ms", summary.durationMs);

            var envNode = node.putObject("environment");
            envNode.put("os", environment.os);
            envNode.put("arch", environment.arch);
            envNode.put("binary_version", environment.binaryVersion);
            envNode.put("runtime_version", environment.runtimeVersion);

            return node;
        }
    }

    static class Summary {
        int total;
        int passed;
        int failed;
        int skipped;
        int errors;
        long durationMs;
    }

    static class Environment {
        String os;
        String arch;
        String binaryVersion;
        String runtimeVersion;
    }

    private boolean compareWithTolerance(double actual, double expected, JsonNode tolerance) {
        if (tolerance == null || !tolerance.isObject()) {
            return Math.abs(actual - expected) < 1e-9;
        }

        if (tolerance.has("abs")) {
            double absTol = tolerance.get("abs").asDouble();
            if (Math.abs(actual - expected) <= absTol) {
                return true;
            }
        }

        if (tolerance.has("rel")) {
            double relTol = tolerance.get("rel").asDouble();
            double diff = Math.abs(actual - expected);
            double avg = (actual + expected) / 2.0;
            if (avg > 0.0 && diff / avg <= relTol) {
                return true;
            }
        }

        return false;
    }

    private JsonNode findTolerance(JsonNode tolerances, String path) {
        if (tolerances == null || !tolerances.isObject()) {
            return null;
        }

        if (tolerances.has(path)) {
            return tolerances.get(path);
        }

        Iterator<String> keys = tolerances.fieldNames();
        while (keys.hasNext()) {
            String key = keys.next();
            if (key.contains("*")) {
                String pattern = key.replace("*", ".*");
                if (path.matches(pattern)) {
                    return tolerances.get(key);
                }
            }
        }

        return null;
    }

    private boolean[] compareResults(JsonNode actual, JsonNode expected, JsonNode tolerances, String path) {
        // Returns [passed, hasReason]
        if (expected.isObject()) {
            if (actual.isNumber()) {
                double actVal = actual.asDouble();
                if (expected.has("min")) {
                    double min = expected.get("min").asDouble();
                    if (actVal < min) {
                        return new boolean[]{false, true};
                    }
                }
                if (expected.has("max")) {
                    double max = expected.get("max").asDouble();
                    if (actVal > max) {
                        return new boolean[]{false, true};
                    }
                }
                if (expected.has("value")) {
                    double expVal = expected.get("value").asDouble();
                    JsonNode tol = findTolerance(tolerances, path);
                    if (!compareWithTolerance(actVal, expVal, tol)) {
                        return new boolean[]{false, true};
                    }
                }
            } else if (actual.isTextual()) {
                String actStr = actual.asText();
                if (expected.has("min_length")) {
                    int minLen = expected.get("min_length").asInt();
                    if (actStr.length() < minLen) {
                        return new boolean[]{false, true};
                    }
                }
                if (expected.has("contains")) {
                    JsonNode contains = expected.get("contains");
                    if (contains.isArray()) {
                        for (JsonNode item : contains) {
                            if (!actStr.contains(item.asText())) {
                                return new boolean[]{false, true};
                            }
                        }
                    }
                }
            } else if (actual.isArray()) {
                int actLen = actual.size();
                if (expected.has("min")) {
                    int min = expected.get("min").asInt();
                    if (actLen < min) {
                        return new boolean[]{false, true};
                    }
                }
                if (expected.has("max")) {
                    int max = expected.get("max").asInt();
                    if (actLen > max) {
                        return new boolean[]{false, true};
                    }
                }
            } else if (actual.isObject()) {
                Iterator<String> fields = expected.fieldNames();
                while (fields.hasNext()) {
                    String key = fields.next();
                    String newPath = path.isEmpty() ? key : path + "." + key;
                    if (!actual.has(key)) {
                        return new boolean[]{false, true};
                    }
                    boolean[] result = compareResults(actual.get(key), expected.get(key), tolerances, newPath);
                    if (!result[0]) {
                        return result;
                    }
                }
            }
        } else if (expected.isArray() && actual.isArray()) {
            for (int i = 0; i < expected.size(); i++) {
                String newPath = path + "[" + i + "]";
                if (i >= actual.size()) {
                    return new boolean[]{false, true};
                }
                boolean[] result = compareResults(actual.get(i), expected.get(i), tolerances, newPath);
                if (!result[0]) {
                    return result;
                }
            }
        } else {
            if (!actual.equals(expected)) {
                return new boolean[]{false, true};
            }
        }
        return new boolean[]{true, false};
    }

    private JsonNode executeMethod(String method, String fixture, JsonNode options) {
        // This is a stub - replace with actual SDK calls when available
        ObjectNode result = mapper.createObjectNode();

        switch (method) {
            case "extract":
                result.put("schema_version", "1.0");
                ObjectNode metadata = result.putObject("metadata");
                metadata.put("page_count", 1);
                break;
            case "extract_text":
                return mapper.getNodeFactory().textNode("Sample text content");
            case "extract_markdown":
                return mapper.getNodeFactory().textNode("# Sample Markdown\n\nContent here");
            case "hash":
                result.put("hash", "abc123");
                result.put("fast_hash", "def456");
                break;
            default:
                break;
        }

        return result;
    }

    private TestResult runTestCase(JsonNode testCase, String schemaVersion, String fixturesBase) {
        long start = System.currentTimeMillis();

        String id = testCase.get("id").asText();

        // Check min_schema_version
        if (testCase.has("min_schema_version")) {
            String minVer = testCase.get("min_schema_version").asText();
            if (compareVersions(schemaVersion, minVer) < 0) {
                TestResult result = new TestResult(id, TestStatus.SKIP, System.currentTimeMillis() - start);
                result.reason = String.format("Schema version %s < minimum required %s", schemaVersion, minVer);
                return result;
            }
        }

        String fixture = testCase.get("fixture").asText();
        String method = testCase.get("method").asText();
        JsonNode options = testCase.get("options");
        JsonNode expected = testCase.get("expected");
        JsonNode tolerances = testCase.has("tolerances") ? testCase.get("tolerances") : null;

        String fixturePath = fixture.startsWith("http") ? fixture : Paths.get(fixturesBase, fixture).toString();

        try {
            JsonNode actual = executeMethod(method, fixturePath, options);
            boolean[] result = compareResults(actual, expected, tolerances, "");

            if (result[0]) {
                TestResult tr = new TestResult(id, TestStatus.PASS, System.currentTimeMillis() - start);
                tr.actual = actual;
                tr.expected = expected;
                return tr;
            } else {
                TestResult tr = new TestResult(id, TestStatus.FAIL, System.currentTimeMillis() - start);
                tr.actual = actual;
                tr.expected = expected;
                tr.reason = "Comparison failed";
                return tr;
            }
        } catch (Exception e) {
            TestResult tr = new TestResult(id, TestStatus.ERROR, System.currentTimeMillis() - start);
            tr.expected = expected;
            tr.error = e.getMessage();
            return tr;
        }
    }

    private int compareVersions(String v1, String v2) {
        String[] parts1 = v1.split("\\.");
        String[] parts2 = v2.split("\\.");

        for (int i = 0; i < Math.min(parts1.length, parts2.length); i++) {
            int n1 = Integer.parseInt(parts1[i]);
            int n2 = Integer.parseInt(parts2[i]);

            if (n1 < n2) return -1;
            if (n1 > n2) return 1;
        }

        return Integer.compare(parts1.length, parts2.length);
    }

    public ConformanceReport runConformance(String suitePath, String outputPath) throws IOException {
        System.out.println("pdftract SDK Conformance Runner");
        System.out.println("SDK: " + SDK_NAME + " v" + SDK_VERSION);
        System.out.println("Suite: " + suitePath);
        System.out.println();

        JsonNode suite = mapper.readTree(new File(suitePath));
        String suiteVersion = suite.get("version").asText();
        String schemaVersion = suite.get("schema_version").asText();
        JsonNode cases = suite.get("cases");

        String fixturesBase = Paths.get(suitePath).getParent().resolve("fixtures").toString();

        System.out.println("Found " + cases.size() + " test cases");
        System.out.println();

        long start = System.currentTimeMillis();
        List<TestResult> results = new ArrayList<>();

        for (JsonNode testCase : cases) {
            TestResult result = runTestCase(testCase, schemaVersion, fixturesBase);

            System.out.println("[" + result.status + "] " + result.id + " (" + result.durationMs + "ms)");

            if (result.status == TestStatus.FAIL || result.status == TestStatus.ERROR) {
                if (result.reason != null) {
                    System.out.println("  Reason: " + result.reason);
                }
                if (result.error != null) {
                    System.out.println("  Error: " + result.error);
                }
            }

            results.add(result);
        }

        long durationMs = System.currentTimeMillis() - start;

        Summary summary = new Summary();
        summary.total = results.size();
        summary.passed = (int) results.stream().filter(r -> r.status == TestStatus.PASS).count();
        summary.failed = (int) results.stream().filter(r -> r.status == TestStatus.FAIL).count();
        summary.skipped = (int) results.stream().filter(r -> r.status == TestStatus.SKIP).count();
        summary.errors = (int) results.stream().filter(r -> r.status == TestStatus.ERROR).count();
        summary.durationMs = durationMs;

        System.out.println();
        System.out.println("Summary:");
        System.out.println("  Total:   " + summary.total);
        System.out.println("  Passed:  " + summary.passed);
        System.out.println("  Failed:  " + summary.failed);
        System.out.println("  Skipped: " + summary.skipped);
        System.out.println("  Errors:  " + summary.errors);
        System.out.println("  Time:    " + summary.durationMs + "ms");

        Environment env = new Environment();
        env.os = System.getProperty("os.name");
        env.arch = System.getProperty("os.arch");
        env.binaryVersion = SDK_VERSION;
        env.runtimeVersion = System.getProperty("java.version");

        ConformanceReport report = new ConformanceReport();
        report.sdk = SDK_NAME;
        report.sdkVersion = SDK_VERSION;
        report.suiteVersion = suiteVersion;
        report.schemaVersion = schemaVersion;
        report.timestamp = Instant.now().atZone(ZoneOffset.UTC).format(DateTimeFormatter.ISO_INSTANT);
        report.results = results;
        report.summary = summary;
        report.environment = env;

        mapper.writerWithDefaultPrettyPrinter().writeValue(new File(outputPath), report.toJson(mapper));

        System.out.println();
        System.out.println("Report written to: " + outputPath);

        return report;
    }

    public static void main(String[] args) throws Exception {
        String suitePath = args.length > 0 ? args[0] : SUITE_PATH;
        String outputPath = args.length > 1 ? args[1] : "conformance-report.json";

        ConformanceTest test = new ConformanceTest();
        ConformanceReport report = test.runConformance(suitePath, outputPath);

        System.exit(report.summary.failed == 0 && report.summary.errors == 0 ? 0 : 1);
    }
}
